File: Usuario/mobile/app/src/main/java/com/signspeak/data/remote/api/ConversacionApi.kt
package com.signspeak.data.remote.api

import com.signspeak.data.remote.dto.conversacion.*
import com.signspeak.data.remote.dto.diccionario.PageResponse
import retrofit2.http.*

interface ConversacionApi {
    @POST("api/conversaciones")
    suspend fun crearConversacion(
        @Body request: CrearConversacionRequest
    ): ConversacionDTO

    @POST("api/conversaciones/{id}/mensajes")
    suspend fun agregarMensaje(
        @Path("id") id: Long,
        @Body request: AgregarMensajeRequest
    ): MensajeConversacionDTO

    @PUT("api/conversaciones/{id}/finalizar")
    suspend fun finalizarConversacion(
        @Path("id") id: Long,
        @Body request: FinalizarConversacionRequest? = null
    ): ConversacionDTO

    @GET("api/conversaciones/{id}")
    suspend fun obtenerConversacion(
        @Path("id") id: Long
    ): ConversacionDTO

    @GET("api/conversaciones/{id}/completa")
    suspend fun obtenerConversacionCompleta(
        @Path("id") id: Long
    ): ConversacionDTO

    @GET("api/conversaciones/usuario/{idUsuario}")
    suspend fun listarConversacionesPorUsuario(
        @Path("idUsuario") idUsuario: Long
    ): List<ConversacionDTO>

    @GET("api/conversaciones/usuario/{idUsuario}/paginado")
    suspend fun listarConversacionesPaginadas(
        @Path("idUsuario") idUsuario: Long,
        @Query("page") page: Int = 0,
        @Query("size") size: Int = 10,
        @Query("sortBy") sortBy: String = "fechaInicio",
        @Query("sortDir") sortDir: String = "DESC"
    ): PageResponse<ConversacionDTO>

    @GET("api/conversaciones/usuario/{idUsuario}/guardadas")
    suspend fun listarConversacionesGuardadas(
        @Path("idUsuario") idUsuario: Long
    ): List<ConversacionDTO>

    @GET("api/conversaciones/usuario/{idUsuario}/buscar")
    suspend fun buscarPorRangoFechas(
        @Path("idUsuario") idUsuario: Long,
        @Query("inicio") inicio: String, // LocalDateTime en formato ISO
        @Query("fin") fin: String // LocalDateTime en formato ISO
    ): List<ConversacionDTO>

    @PUT("api/conversaciones/{id}/guardar")
    suspend fun marcarComoGuardada(
        @Path("id") id: Long,
        @Query("guardada") guardada: Boolean
    ): ConversacionDTO

    @DELETE("api/conversaciones/{id}")
    suspend fun eliminarConversacion(
        @Path("id") id: Long
    )

    @GET("api/conversaciones/{id}/mensajes")
    suspend fun obtenerMensajes(
        @Path("id") id: Long
    ): List<MensajeConversacionDTO>

    @GET("api/conversaciones/usuario/{idUsuario}/count")
    suspend fun contarConversacionesActivas(
        @Path("idUsuario") idUsuario: Long
    ): Long
}

File: Usuario/frontend/src/services/conversacionApi.js
import axios from 'axios';

const API_URL = 'http://localhost:8081/api/conversaciones'; 
class ConversacionAPI {
  
  // Crear nueva conversación
  async crearConversacion(idUsuario, titulo = 'Conversación sin título') {
    try {
      const response = await axios.post(API_URL, {
        idUsuario,
        tituloConversacion: titulo // ✅ Cambiar 'titulo' a 'tituloConversacion'
      });
      console.log('✅ Conversación creada:', response.data);
      return response.data;
    } catch (error) {
      console.error('❌ Error creando conversación:', error);
      throw error;
    }
  }

  // Agregar mensaje a conversación
  async agregarMensaje(idConversacion, mensaje) {
    try {
      // ✅ Cambiar ruta a /{id}/mensajes
      const response = await axios.post(`${API_URL}/${idConversacion}/mensajes`, {
        tipoMensaje: mensaje.tipoMensaje || 'seña_detectada',
        contenidoTexto: mensaje.contenidoTexto || null,
        idSenaDetectada: mensaje.señaDetectada || null, // ✅ Cambiar nombre
        confianzaDeteccion: mensaje.confianzaDeteccion || null
      });
      return response.data;
    } catch (error) {
      console.error('❌ Error agregando mensaje:', error);
      throw error;
    }
  }

  // Guardar conversación (CA-17) ⭐
  async guardarConversacion(idConversacion, guardada = true) {
    try {
      // ✅ Agregar parámetro 'guardada'
      const response = await axios.put(`${API_URL}/${idConversacion}/guardar?guardada=${guardada}`);
      console.log('✅ Conversación guardada:', response.data);
      return response.data;
    } catch (error) {
      console.error('❌ Error guardando conversación:', error);
      throw error;
    }
  }

  // Finalizar conversación
  async finalizarConversacion(idConversacion, resumen = null) {
    try {
      const body = resumen ? { resumenConversacion: resumen } : {};
      const response = await axios.put(`${API_URL}/${idConversacion}/finalizar`, body);
      console.log('✅ Conversación finalizada:', response.data);
      return response.data;
    } catch (error) {
      console.error('❌ Error finalizando conversación:', error);
      throw error;
    }
  }

  // Listar conversaciones del usuario
  async listarConversaciones(idUsuario) {
    try {
      const response = await axios.get(`${API_URL}/usuario/${idUsuario}`);
      return response.data;
    } catch (error) {
      console.error('❌ Error listando conversaciones:', error);
      throw error;
    }
  }

  // Listar solo conversaciones guardadas
  async listarConversacionesGuardadas(idUsuario) {
    try {
      const response = await axios.get(`${API_URL}/usuario/${idUsuario}/guardadas`);
      return response.data;
    } catch (error) {
      console.error('❌ Error listando conversaciones guardadas:', error);
      throw error;
    }
  }

  // Obtener conversación específica con todos sus mensajes
  async obtenerConversacion(idConversacion) {
    try {
      // ✅ Usar /completa para obtener con mensajes
      const response = await axios.get(`${API_URL}/${idConversacion}/completa`);
      return response.data;
    } catch (error) {
      console.error('❌ Error obteniendo conversación:', error);
      throw error;
    }
  }

  // Obtener solo los mensajes de una conversación
  async obtenerMensajes(idConversacion) {
    try {
      const response = await axios.get(`${API_URL}/${idConversacion}/mensajes`);
      return response.data;
    } catch (error) {
      console.error('❌ Error obteniendo mensajes:', error);
      throw error;
    }
  }

  // Eliminar conversación
  async eliminarConversacion(idConversacion) {
    try {
      await axios.delete(`${API_URL}/${idConversacion}`);
      console.log('✅ Conversación eliminada');
      return true;
    } catch (error) {
      console.error('❌ Error eliminando conversación:', error);
      throw error;
    }
  }

  // Contar conversaciones activas
  async contarConversacionesActivas(idUsuario) {
    try {
      const response = await axios.get(`${API_URL}/usuario/${idUsuario}/count`);
      return response.data;
    } catch (error) {
      console.error('❌ Error contando conversaciones:', error);
      throw error;
    }
  }
}

export default new ConversacionAPI();